dra-evolution/pkg/api/capacity_types.go

```diff
@@ -1,170 +1,205 @@
-	SemVerValue   *SemVer            `json:"semVerValue,omitempty"`
+	// DeviceResourceValue is an embedded type representing the actual value of the device resource.
+	DeviceResourceValue `json:",inline" protobuf:"bytes,2,opt,name=value"`
 }
 
-func (a Attribute) Equal(b Attribute) bool {
-	if a.Name != b.Name {
-		return false
-	}
+// DeviceResourceValue represents the value of a device resource.
+// DeviceResourceValue must have one and only one field set.
+type DeviceResourceValue struct {
+	// QuantityValue is a quantity.
+	QuantityValue *resource.Quantity `json:"quantity,omitempty" protobuf:"bytes,1,opt,name=quantity"`
 
-	return a.EqualValue(b)
+	// IntRangeValue is a range of 64-bit integers.
+	IntRangeValue *IntRange `json:"intRange,omitempty" protobuf:"varint,2,rep,name=intRange"`
 }
 
-func (a Attribute) EqualValue(b Attribute) bool {
-	if a.StringValue != nil && b.StringValue != nil && *a.StringValue == *b.StringValue {
-		return true
-	}
+// DeviceResourceSources represents a set of DeviceResources from many possible sources.
+type DeviceResourceSources struct {
+	// FromSharedPool holds the list of DeviceResources that are consumed from the shared pool.
+	FromSharedPool []DeviceResource `json:"fromSharedPool,omitempty" protobuf:"bytes,1,opt,name=fromSharedPool"`
+
+	// Local holds the list of DeviceResources that are local to just this device.
+	Local []DeviceResource `json:"local,omitempty" protobuf:"bytes,2,opt,name=local"`
+}
 
-	if a.IntValue != nil && b.IntValue != nil && *a.IntValue == *b.IntValue {
-		return true
-	}
+// DeviceAttribute is a combination of an attribute name and its value.
+type DeviceAttribute struct {
+	// Name is a unique identifier across all possible attributes of devices.
+	//
+	// If this is a DNS subdomain (no dot), then the driver name gets added
+	// when looking up attributes. This avoids name collisions with attributes
+	// used by other drivers.
+	//
+	// If this is a full DNS domain, then the meaning of the attribute driver-independent.
+	// For example, Kubernetes will use `*.k8s.io` names when defining attributes that
+	// drivers from different vendors are supposed to use.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
 
-	if a.QuantityValue != nil && b.QuantityValue != nil && (*a.QuantityValue).Equal(*b.QuantityValue) {
-		return true
-	}
+	// DeviceAttributeValue is an embedded type representing the actual value of the device attribute.
+	DeviceAttributeValue `json:",inline" protobuf:"bytes,2,opt,name=attributeValue"`
+}
 
-	if a.SemVerValue != nil && b.SemVerValue != nil && *a.SemVerValue == *b.SemVerValue {
-		return true
-	}
+// The Go field names below have a Value suffix to avoid a conflict between the
+// field "String" and the corresponding method. That method is required.
+// The Kubernetes API is defined without that suffix to keep it more natural.
+
+// DeviceAttributeValue must have one and only one field set.
+type DeviceAttributeValue struct {
```

Comment:
K8s CEL doesn't have a semantic version type. I added that to the DRA CEL environment for 1.30 because it's needed for the "device driver version check" use case.

@klueska also had some usage for int slice and string slice, so I added those, too. I don't know how strongly he feels about removing them?

Reply:
IntSlice has become IntRange and has moved out of being an attribute to being a `DeviceResource` type.
(meaning I don't feel strongly about removing IntSlice).

StringSlice is useful for providing a set of related attributes and checking for "membership" in that set. In theory that could be accomplished with existence checks on a distinct attribute though, so I don't feel too strongly.

I definitly need the ability to do semver comparisons. I'd be OK if these were just stored as raw strings though, so long as we have access to a method for performing semver comparisons on those strings.

Reply:
> StringSlice is useful for providing a set of related attributes and checking for "membership" in that set.

Can you explain the actual use-case for this?  Is it something like:

- Device has N "feature extensions", published as `features: [ foo, bar, bat, qux ]`
- Claim wants to express "features[] has foo" 
?

Would it be simpler to do:
- Device has `featureFoo:true`, `featureBat:true`, etc
- Claim requests `featureFoo == true` 
?

Reply:
I don't know about "simpler", but it is an option, and why i don't feel too strongly about keeping `StringSlice`.

Reply:
Unless I hear otherwise, I consider `StringSlice`, `IntSlice`, and `Int` candidates for removal from the next update of this PR.

Reply:
This comment can be resolved :)